### PR TITLE
Feature: Better Bookmarks Searching

### PR DESCRIPTION
**Pull request by JoshAshby**
This implements a nicer interface around searching bookmarks, allowing you to filter them down by a wide range of different fields: title, description, url, url host and tags by default, while providing a nice interface to add more fields seamlessly.

screencast 2018-06-08 16-38-42

### CODE REVIEW

File: lib/webpage_cache_service/cache.rb [hunk 1]
```diff
@@ -144,23 +141,5 @@ def build_metadata response:
         headers: response.headers.to_h
       }
     end
-
```
Comment (JoshAshby) [inline lib/webpage_cache_service/cache.rb]: I don't remember doing this ... I should back this file out of the changes

File: app/controllers/bookmarks/search_controller.rb [hunk 2]
```diff
@@ -2,18 +2,38 @@ class Bookmarks::SearchController < ApplicationController
...
+            :error_messages
+          ]
+        )
+    end
 
     respond_to do |format|
       format.html { render :index }
-      format.json { render :index, status: :ok }
+      if @searcher.errors.any?
+        format.json { render :error, status: 500 }
```
Comment (robacarp) [inline app/controllers/bookmarks/search_controller.rb]: should this really be a 500?

File: app/policies/bookmark_searcher_policy.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class BookmarkSearcherPolicy < ApplicationPolicy
+  class Scope < Scope
+    def resolve
+      scope.new.scope({ term: { user_id: user.id } })
```
Comment (robacarp) [inline app/policies/bookmark_searcher_policy.rb]: nice

File: app/searchers/application_searcher.rb [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  end
+
+  protected
+
+  def index_klass
+    self.class.index_klass
+  end
+
+  def keywords_data
+    self.class.keywords_data
```
Comment (robacarp) [inline app/searchers/application_searcher.rb]: this feels like a bunch of misdirection

File: app/searchers/application_searcher.rb [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  end
+
+  def keywords
+    self.class.keywords ||= keywords_data.flat_map do |keyword_data|
+      keyword = keyword_data.keys.first
+      data = keyword_data[ keyword ]
+
+      next as if data[:as].present?
```
Comment (robacarp) [inline app/searchers/application_searcher.rb]: is `as` relevant in this scope?

File: app/searchers/application_searcher.rb [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                                           memo[ field ] = properties[ field ][:type].to_sym
+                                         end
+                                       end
+  end
+
+  def parser_klass
+    self.class.parser_klass ||= begin
+                                  closure_keywords = keywords
+                                  Class.new ApplicationSearcher::BaseQueryParser do
+                                    fields(*closure_keywords)
```
Comment (robacarp) [inline app/searchers/application_searcher.rb]: 🤔

File: app/controllers/bookmarks/search_controller.rb [hunk 7]
```diff
@@ -2,18 +2,38 @@ class Bookmarks::SearchController < ApplicationController
...
-      }
-    ).objects.page params[:page]
+    @query = params[:q]
+    @searcher = policy_scope(BookmarkSearcher)
+
+    if @query.blank? || @query.empty?
+      @bookmarks = Bookmark.none
+    else
+      if @query
+        odd_quotes = @query.count('"').odd?
```
Comment (robacarp) [inline app/controllers/bookmarks/search_controller.rb]: no love for the single quote?

Comment (JoshAshby) [reply app/controllers/bookmarks/search_controller.rb]: Nope, figured it was going to be not too fun with apostrophes that I'd rather just not deal with them for now until I see how well this works

File: app/controllers/bookmarks/search_controller.rb (hunk 2 above)
Comment (JoshAshby) [reply app/controllers/bookmarks/search_controller.rb]: Hmm, probably not 🤷‍♀️ I need to figure out a better way to surface errors with the built-in [semantic-ui search component](https://semantic-ui.com/modules/search.html) though

File: app/searchers/application_searcher.rb (hunk 4 above)
Comment (JoshAshby) [reply app/searchers/application_searcher.rb]: the name or the class attribute accessors inside the instance?

File: app/searchers/application_searcher.rb (hunk 5 above)
Comment (JoshAshby) [reply app/searchers/application_searcher.rb]: Left over, but means I could do like
```ruby
keyword :superlong_field, as: :short
```
But I haven't ended up needing it yet, so I did consider removing this little bit

File: app/searchers/application_searcher.rb (hunk 6 above)
Comment (JoshAshby) [reply app/searchers/application_searcher.rb]: dude

File: app/searchers/application_searcher/elasticsearch_query.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+      return { match_phrase: { @clause.field => @clause.phrase.to_s } } if @clause.is_a? PhraseClause
+      { match: { @clause.field => @clause.term.to_s } }
+    end
+  end
+
+  class ElasticsearchQuery
+    attr_reader :query
+
+    HEURISTICS_MAP = {
```
Comment (JoshAshby) [inline app/searchers/application_searcher/elasticsearch_query.rb]: @robacarp not super sure about this bit, but I can't find much of a cleaner way to do this all :/

**Comment by JoshAshby** [comment]
current implementation demo:

still not 100% on it but its better than whats in there now

**Pull request closed without merging**